keps/sig-cli/kubectl-apply-prune-beta.md

```diff
@@ -0,0 +1,250 @@
+
+## Proposal
+
+### Challenges
+The challenges of pruning are mainly from two aspects.
+- Find out correct objects to delete. When a directory is applied after some updates, it loses the track of objects that are previously applied from the same directory.
+- Delete objects at proper time. Assume the set of objects to be deleted is available, when to delete them is another challenge. Those objects could be still needed during rolling out other objects. One example is a ConfigMap object being referred in a Deployment object. When rolling update happens, new pods refer to new ConfigMap and old pods are still running and referring to old ConfigMap. The old ConfigMap can't be deleted until the rolling up successfully finishes.
+
+In this design, we focus on a solution for the first challenge. For the second challenge, we can blacklist certain types such as ConfigMaps and Secrets for deletion. Generally deleting objects at proper time can be better addressed by Garbage Collection.
+
+### limitation of Alpha prune
+
+The Alpha prune works by label selector:
+```
+  # Apply the configuration in manifest.yaml that matches label app=nginx and delete all the other resources that are not in the file and match label app=nginx.
+  kubectl apply --prune -f manifest.yaml -l app=nginx
+```
+Internally, `kubectl apply` takes following logic to perform pruning.
+- For one namespace, it queries all types of resources except [some hard coded ones](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply/apply.go#L632)
+to find objects that matches the label. The number of queries within one namespace is about the number of API resources installed in the cluster.
+- Among the matched objects, skip the ones without [last-applied-configuration annotation](https://kubernetes.io/docs/concepts/overview/object-management-kubectl/declarative-config/#how-apply-calculates-differences-and-merges-changes).
+- Among the rest objects, delete any object that is not present in the new apply set.
+- Repeat this for all visited namespaces. Overall, the number of queries is propotional to the number of API resources installed in the cluster.
+
+The limitations are mainly from three aspects:
+- Objects might be deleted unintentionally when different directories use the same labels
+- Some types are hard-coded to avoid being pruned.
+- The query number is neither constant nor linear to the number of resources being applied/pruned. 
+
+### Apply-Config Object
+To solve the first challenge, a concept of apply-config object is proposed. An apply-config object is an object that contains the GKN information for all the objects that are to be applied. The apply-config object is not generated by kubectl when applying is running. Instead, it should be generated by other tools such as kustomize, Helm. When generating resources to be consumed by `kubectl apply`, those tools are responsible to generate the associated apply-config object to utilize better pruning.
```

Comment:
I'm curious why Kustomize/Helm has to generate this GKN rather than the `kubectl apply` APIs itself. Wouldn't there be interoperability concerns between tools generating differing apply-config objects?

Helm currently manages the last applied objects within an object called a "Release", which contains the manifests last applied. in a sense, we've already solved the pruning problem for Helm, and this is trying to solve the same problem. By switching over to the Apply-Config object, what benefits would tools like Helm gain? I'm trying to gauge here what benefits would a Helm user gain if we switched the state machine internals to use the apply-config object instead for tracking what objects need to be pruned.

Reply:
On one side, `kubectl apply` may not get the correct list of objects. Users may change directories when running successive applies. Different directories will lead `kubectl` generate different lists of objects. In addition, `kubectl` is lack of the ability of generating a unique identifier, whether it is apply-cfg object with a unique name or a unique label. `kubectl` itself doesn't handle it. On the other side, there are tools that can be combined to use with `kubectl`. The tools generate resources and kubectl apply whatever is piped to it. So generating a apply-cfg object associated with the list of resource avoids makes sense.